### PR TITLE
[nova] Shellinabox & Dell KMS Proxy in Nginx

### PR DESCRIPTION
**Pull request by fwiesel**
This implements the shellinabox proxy in nginx
as lua scripts, and also solves proxying kvm
for dell hosts.

The existing custom solution in nova only proxies
to a single host, and cannot be changed easily
to multiple.

For the Dell KMS, it also needs to
proxy websockets, which is simply a configuration
option in nginx, but would otherwise need
to be implemented in the python proxy

### CODE REVIEW

**Comment by fwiesel** [comment]
This goes together with the change in nova sapcc/nova#379

File: openstack/nova/shellinabox/revproxy.lua [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+    local new_location = token_base .. path .. "?" .. new_query
+    ngx.header.Location = new_location
+end
+
+function _M.process()
+    local entry = _M.lookup()
+    if not entry then return end
+    local host = host_from_entry(entry)
+    local request_uri = ngx.var.request_uri
+    local sub_uri = request_uri:sub(38)
```
Comment (joker-at-work) [inline openstack/nova/shellinabox/revproxy.lua:212]: Can you point me to any docs for this `sub()`? I can only find `string.sub()` and would like to find out what the `38` might mean here.

File: openstack/nova/shellinabox/revproxy.lua [hunk 2]
```diff
@@ -0,0 +1,226 @@
+local _M = {}
+local cjson = require "cjson"
+local config = require "config"
+local resty_sha256 = require "resty.sha256"
+local str = require "resty.string"
+local memcached = require "resty.memcached"
+local mysql = require "resty.mysql"
+local mlcache_loaded, mlcache = pcall(require, "resty.mlcache")
+local x = "%x"
+local t = { x:rep(8), x:rep(4), x:rep(4), x:rep(4), x:rep(12) }
```
Comment (joker-at-work) [inline openstack/nova/shellinabox/revproxy.lua:10]: What that do? Is that string-formatting for numbers?

**Review by joker-at-work — COMMENTED** [review]
Do you have some hint for a lua/openresty tutorial I could read so this review of mine could make any form of sense?

File: openstack/nova/shellinabox/revproxy.lua [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+    return t1
+end
+
+local function host_from_entry(entry)
+    if entry.port and entry.port ~= 443 then
+        return "https://" .. entry.host .. ":" .. entry.port
+    else
+        return "https://" .. entry.host
+    end
+end
```
[the remark this answers is not in the document]
Comment (fwiesel) [reply openstack/nova/shellinabox/revproxy.lua:61]: Yes, the code would be simpler. I find it though more difficult to parse when reading the logs.

File: openstack/nova/shellinabox/revproxy.lua (hunk 2 above)
Comment (fwiesel) [reply openstack/nova/shellinabox/revproxy.lua:10]: Well, yes and no. It is the same "code", but for matching [patterns](https://www.lua.org/manual/5.3/manual.html#6.4.1). They are similar to regular expressions, but less powerful/complex.
The alternative would be hard-coding the "%x%x%x..-%x%x"... not sure if that is easier to read. I was fearing, I might miss a `%x` somewhere.

Comment (joker-at-work) [reply openstack/nova/shellinabox/revproxy.lua:10]: Ah. It means "repeat". Thanks.

**Comment by fwiesel** [comment]
The canonical introduction would be [Programming in Lua](https://www.lua.org/pil/contents.html).
Personally, I didn't read it cover from cover, but I often end up there looking up some basics again.
Which happed when writing this too, as I also do not remember how the functions are called here again

I do not claim to have mastered the language, but I think the more difficult part is rather the nginx library.
I look most thinks up in the [nginx-api for lua](https://github.com/openresty/lua-nginx-module#nginx-api-for-lua) 
Or about  [nginx request phases](https://nginxguts.com/2011/01/phases.html)

Coming from python, the language shouldn't be to hard. Both come from an educative background as imperative with object orientation added after the fact. Especially at the scope we are looking (no inheritance, metatables...)

In python the map doing the object orientation is hidden as `__dict__` , and there are some magic methods which handle the member/attribute resolution  (`__getattr__`, `__getattribute__`).  That is handled in lua by "syntactic sugar".

A very short introduction form my side:

`x.y`  ⇔ `x["y"]`
`x:y(...)` ⇔ `x.y(x, ...)`
`function a(...)`  ⇔ `a = function(...)`, similar to python (or javascript) where the function declaration returns the function itself
So in summary: `function a.b(x, ...)`  ⇔ `a["b"] = function(x, ...)`
`[[...]]` ⇔ `"..."`
`..`  concatenate objects as string (`nil` doesn't work though)
`...` varargs in lua, which can be passed on to other functions easily.
So, this is actual code: `function a(...) print(...) end` and simply passes all arguments passed to `a` over to `print`. 

In lua there is no `@` for decorators, so I called them explicitly.

All variables should be declared `local` because like in javascript, they went with a global default.

Similar to python, when a module is required, the code is executed once the first time it is imported.
But in contrast to python there is no implicit namespace where the file lies. 
The namespaces result from the fact, that `require` returns what the module returns when it is executed, and there you create a table, which contains what you want to be exported.
By declaring objects `local` and not explicitly adding them to the returned table they become `private` in other languages.

Like in python, `"a.b"` in a module translates to `a/b` in the path.

File: openstack/nova/shellinabox/revproxy.lua (hunk 1 above)
Comment (fwiesel) [reply openstack/nova/shellinabox/revproxy.lua]: I've added some more comments, I hope that clarifies things a bit more

**Pull request merged**